### ISSUE TITLE
Wrapping text around an image

### ISSUE THREAD

**Issue by badgermaniac** [naturalcrit/homebrewery#380]
First, thanks for the awesome tool. 

I have been inserting images into my docs and positioning them in various ways. However, I haven't figured out how to wrap text around an image.  

I have cut out my images so they have transparent backgrounds, leaving the border irregular as oppose to rectangular, but now I want the text to "fill in the gaps" (see pg. 125 in PHB). 

Any code help would be appreciated. Thanks.

### Additional Details

**Share Link** :

or

**Brew code to reproduce** : <details><summary>Click to expand</summary><code><pre>

PASTE BREW CODE HERE

</pre></code></details>

**Comment by BaronMurph** [comment]
Haven't actually figured out how to add multiple spaces (or, like, a horizontal in-line spacer) myself, so I'll admit to leaving a comment here in part for selfish reasons in case anyone does. 

What I _can_ tell you, however, is that you can add line breaks without spaces in front (along the lines of https://i0.wp.com/www.critical-hits.com/wp-content/uploads/2014/08/Bulette.jpg), just by adding a double space at the end of a line.
`So just like this with two spaces over there -->  `
`And then this show up without the customary spaces in front. Note that this is on the next line, not two lines down - I only pressed the Enter key once, so to speak, not twice.`

**Comment by badgermaniac** [comment]
Thanks. This helps for aligning images on one side of text, so it is s start.

**Comment by BaronMurph** [comment]
Hell yeah, found it.
&nbsp; adds a single space, and you can just keep adding 'em. It'll look ugly on code ('cause you need to keep adding it, so two spaces would be &nbsp;&nbsp; and three &nbsp;&nbsp;&nbsp;) but it works.
>&nbsp;&nbsp;&nbsp; like this

**Comment by Biotronic** [comment]
The solution you really want for this is CSS's [shape-outside](https://css-tricks.com/almanac/properties/s/shape-outside/). Here's an example of how to do it:

http://homebrewery.naturalcrit.com/share/Hkgyq2eJ-

**Comment by Biotronic** [comment]
This report could be turned into an enhancement with a bit of javascript and a CSS class:

```
$(".wrap-image").each(function() {
    if ($(this).attr("src")) {
        $(this).css("shape-outside", "url("+$(this).attr("src")+")"); 
    }
});
```

This will set the 'shape image' of any image to its image, and thus automatically do what I did in the previous post.

**Comment by calculuschild** [comment]
This might be a candidate for a snippet... or possibly added to the FAQ.

**Comment by ericscheid** [comment]
Linking to FAQ issue #810

**Comment by ericscheid** [comment]
Tagging as #help-wanted to get a tute/faq written, untagging as #feature.

**Comment by ericscheid** [comment]
Just in case that example brew by @Biotronic goes away or gets broken, this is what the source brew is:

```
<img src="https://i.imgur.com/MloXv1M.png" style="shape-outside: url('https://i.imgur.com/MloXv1M.png'); float: right;">
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
```

The pertinent bit is adding `style="shape-outside: url('the-image-url'); float: right;"` inside the `<img ..>` tag.

The effect can also be done with `<style>...</style>` and some clever selector targeting.

**Comment by Biotronic** [comment]
Using `<style>` would be much nicer, if I understand correctly. My immediate idea was to use `shape-outside: attr(src);`, but `attr()` [is not very well supported](https://developer.mozilla.org/en-US/docs/Web/CSS/attr()#browser_compatibility). In a perfect world, this would Just Work™:
```css
  .wrap-image-left {
    shape-outside: attr(src);
    float: left;
  }
  .wrap-image-right {
    shape-outside: attr(src);
    float: right;
  }
```

**Comment by Fionn-sparse** [comment]
> Linking to FAQ issue [#810](https://github.com/naturalcrit/homebrewery/issues/810)





https://tendrildavion.github.io/jacques778/

**Comment by G-Ambatte** [comment]
> > Linking to FAQ issue [#810](https://github.com/naturalcrit/homebrewery/issues/810)
> 
> https://tendrildavion.github.io/jacques778/

Given that the wrapping needs to live in the user space, we can't define it into a grid or flex layout. What we did was modify the Markdown parser to automatically copy the `src` property into a variable - as I recall, `HB_src`. We then added new styles for `wrapLeft` and `wrapRight` that adds properties like `float` and `shape-outside`, using this `HB_src` variable.

It would have been better if we could have used `attr(src)` but this will work until the browser implementation catches up.
